File: hopelink backend/backend/src/models/organization.model.js
import mongoose from "mongoose";

const fileSchema = new mongoose.Schema({
  url: { type: String, required: true },
  publicId: { type: String },
  originalName: { type: String },
  mimeType: { type: String },
  size: { type: Number },
  uploadedAt: { type: Date, default: Date.now },
});

const bankDetailsSchema = new mongoose.Schema({
  bankName: { type: String, required: true },
  accountHolderName: { type: String, required: true },
  accountNumber: { type: String, required: true },
  bankBranch: { type: String },
  voidCheque: fileSchema, // nested document
});

const socialMediaSchema = new mongoose.Schema({
  facebook: String,
  instagram: String,
  linkedin: String,
});

const organizationSchema = new mongoose.Schema(
  {
    organizationName: { type: String, required: true },
    organizationType: { type: String, required: true },
    registrationNumber: { type: String, required: true },
    dateOfRegistration: { type: Date },
    officialEmail: { type: String },
    officialPhone: { type: String },
    website: { type: String },
    country: { type: String },
    city: { type: String },
    registeredAddress: { type: String },
    representativeName: { type: String },
    designation: { type: String },
    primaryCause: { type: String },
    missionStatement: { type: String },
    activeMembers: { type: Number, required: true },
    recentCampaigns: [{ type: String }],
    logo: fileSchema,
    registrationCertificate: fileSchema,
    taxCertificate: fileSchema,
    constitutionFile: fileSchema,
    proofOfAddress: fileSchema,
    bankDetails: bankDetailsSchema,
    socialMedia: socialMediaSchema,
    status: { 
      type: String, 
      default: "pending",
      enum: ['pending', 'approved', 'rejected', 'suspended']
    },
    user: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: false
    },
    approvedBy: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
      required: false
    },
    approvedAt: {
      type: Date,
      required: false
    },
    rejectionReason: {
      type: String,
      required: false
    },
    isVerified: { type: Boolean, default: false },
  },
  { timestamps: true }
);

// Check if the model has already been registered
let Organization;
if (mongoose.models.Organization) {
  Organization = mongoose.model('Organization');
} else {
  Organization = mongoose.model('Organization', organizationSchema);
}

export default Organization;

File: hopelink_admin/lib/features/Auth/models/organization_model.dart
// ─────────────────────────────────────────────────────────────
//  MODEL  —  organization_model.dart
// ─────────────────────────────────────────────────────────────

class OrganizationRegistrationRequest {
  // ── Basic Info ──────────────────────────────────────────────
  final String organizationName;
  final String organizationType;
  final String registrationNumber;
  final String dateOfRegistration; // "YYYY-MM-DD"
  final String country;
  final String city;
  final String registeredAddress;

  // ── Contact ─────────────────────────────────────────────────
  final String officialEmail;
  final String officialPhone;
  final String? website;

  // ── Social ──────────────────────────────────────────────────
  final String? facebook;
  final String? instagram;
  final String? linkedin;

  // ── Representative ──────────────────────────────────────────
  final String representativeName;
  final String designation;

  // ── Bank ────────────────────────────────────────────────────
  final String bankName;
  final String accountHolderName;
  final String accountNumber;
  final String bankBranch;

  // ── Mission ─────────────────────────────────────────────────
  final String primaryCause;
  final String missionStatement;
  final int activeMembers;
  final String recentCampaigns;

  // ── Files ───────────────────────────────────────────────────
  final String taxCertificatePath;
  final String constitutionFilePath;
  final String proofOfAddressPath;
  final String voidChequePath;

  const OrganizationRegistrationRequest({
    required this.organizationName,
    required this.organizationType,
    required this.registrationNumber,
    required this.dateOfRegistration,
    required this.country,
    required this.city,
    required this.registeredAddress,
    required this.officialEmail,
    required this.officialPhone,
    this.website,
    this.facebook,
    this.instagram,
    this.linkedin,
    required this.representativeName,
    required this.designation,
    required this.bankName,
    required this.accountHolderName,
    required this.accountNumber,
    required this.bankBranch,
    required this.primaryCause,
    required this.missionStatement,
    required this.activeMembers,
    required this.recentCampaigns,
    required this.taxCertificatePath,
    required this.constitutionFilePath,
    required this.proofOfAddressPath,
    required this.voidChequePath,
  });
}

// ── Success Response ─────────────────────────────────────────
class OrganizationRegistrationResponse {
  final bool success;
  final String message;
  final OrganizationData data;

  const OrganizationRegistrationResponse({
    required this.success,
    required this.message,
    required this.data,
  });

  factory OrganizationRegistrationResponse.fromJson(Map<String, dynamic> json) {
    return OrganizationRegistrationResponse(
      success: json['success'] as bool,
      message: json['message'] as String,
      data: OrganizationData.fromJson(json['data'] as Map<String, dynamic>),
    );
  }
}

class OrganizationData {
  final String id;
  final String name;
  final String status;

  const OrganizationData({
    required this.id,
    required this.name,
    required this.status,
  });

  factory OrganizationData.fromJson(Map<String, dynamic> json) {
    return OrganizationData(
      id: json['id'] as String,
      name: json['name'] as String,
      status: json['status'] as String,
    );
  }
}

// ── Org Type Enum ────────────────────────────────────────────
enum OrganizationType {
  ngo('NGO'),
  ingo('INGO'),
  nonprofit('Non-Profit'),
  charity('Charity'),
  foundation('Foundation'),
  socialEnterprise('Social Enterprise');

  final String label;
  const OrganizationType(this.label);
}
